src/coreComponents/mesh/generators/HexCellBlockManager.cpp

```diff
@@ -0,0 +1,1126 @@
+ * TODO For full tetrahedral meshes we need a dedicated implementation
+ * 
+ * Options : Store if the face is a triangle or quad ? 
+ *  Store a  NO_FACE  value if  same storage space for hybrid meshes
+ *  
+ */
+class MeshConnectivityBuilder
+{
+public:
+  MeshConnectivityBuilder(CellBlockManagerBase & cellBlockManager);
+  MeshConnectivityBuilder( const MeshConnectivityBuilder & ) = delete;
+  MeshConnectivityBuilder & operator=( const MeshConnectivityBuilder & ) = delete;
+  virtual ~MeshConnectivityBuilder() = default;
+
+  localIndex numEdges() const {
+    return nbEdges;
+  }
+  localIndex numFaces() const {
+    return nbFaces;
+  }
+
+  // Dependent on Cell Type
+  virtual bool computeFaces() = 0;
+  virtual bool computeEdges() = 0;
+
+  // Independent on Cell Type
+  bool computeNodesToEdges( ArrayOfSets<localIndex> & result ) const; 
+  bool computeEdgesToNodes( array2d<localIndex>& result ) const;  
+  bool computeNodesToElements( ArrayOfArrays<localIndex>  & result ) const;
+
+  // Dependent on Cell Type
+  virtual bool computeNodesToFaces( ArrayOfSets<localIndex> & result ) const = 0 ;
+  virtual bool computeEdgesToFaces( ArrayOfSets<localIndex> & result ) const  = 0;
+  virtual bool computeFacesToNodes( ArrayOfArrays<localIndex> & result ) const = 0;
+  virtual bool computeFacesToElements( array2d<localIndex> &  result ) const  = 0;
+  virtual bool computeFacesToEdges( ArrayOfArrays<localIndex> & result ) const =0 ;
+
+  // Why is this here? The information is on the CellBlocks
+  virtual bool computeElementsToFacesOfCellBlocks() = 0; 
+  virtual bool computeElementsToEdgesOfCellBlocks() = 0;
+
+  bool debuggingComputeAllMaps() const;
+
+  void printDebugInformation() const;
+
+protected:
+  std::pair< CellBlockIndex, LocalCellIndex> 
+  getBlockCellFromManagerCell( LocalCellIndex cellId ) const;
+
+  unsigned int nbCellBlocks() const {
+    return m_cellBlocks.size();
+  } 
+  CellBlock const & getCellBlock( unsigned int id ) const {
+    return *(m_cellBlocks[id]);
+  }
+  CellBlock & getCellBlock( unsigned int id ){
+    return *(m_cellBlocks[id]);
+  }
+ 
+  void computeAllFacesToUniqueFace( std::vector< localIndex > & result ) const;
+
+  virtual void getFacesAroundEdge( int edgeIndex,
+                           std::vector< localIndex > const & allFacesToUniqueFace, 
+                           std::set <localIndex> & faces ) const = 0;
+
+protected:
+/// Number of vertices 
+SizeOfStuff nbNodes = 0;
+/// All Elements in all CellBlocks
+SizeOfStuff nbElements = 0;  
+
+/// Number of unique edges - Size of m_uniqueEdges 
+SizeOfStuff nbEdges = 0;
+/// Number of unique faces - Size of m_uniqueFaces and m_isBoundaryFace
+SizeOfStuff nbFaces = 0;
+
+/// Cell Blocks on which the class operates - Size of nbBlocks
+std::vector< CellBlock * > m_cellBlocks;
+
+/// Offset for the numbering of all the cells - First value is the number of cells of block 0
+/// Size of nbBlocks
+std::vector< LocalCellIndex > m_blockCellIndexOffset;
```

Comment:
Do you have to take into account an MPI offset too?

Reply:
I do not think so. If I got it right, the CellBlockManager is operating on one MPI rank.